File: backend/services/game.py
from fastapi import WebSocket
import asyncio

class Game:
    def __init__(self):
        self.players = [] # list [WebSocket]
        self.board = [["" for _ in range(3)] for _ in range(3)]
        self.isRunning = True
        self.currentPlayer = "X"

    async def connect(self, ws: WebSocket):
        await ws.accept()
        if (len(self.players) >= 2):
            await ws.send_json({
                "type": "rejection",
                "content": "party is full"
            })
            await ws.close()
            return False

        availableChar = self.availableChar() 
        self.players.append({
            "ws": ws,
            "char": availableChar
        })


        await ws.send_json({
            "type": "meInfo",
            "content": availableChar
        })

        await self.updateGameState()

        print("Player Connected")
        print(f"number of players {len(self.players)}")

        return True

    async def handleEvenet(self, event: dict, ws: WebSocket):
        if event["type"] == "play":

            if not self.isRunning:
                await self.resetGameState()

            playerChar = self.getPlayerFromWebSocket(ws)
            if playerChar != self.currentPlayer:
                await ws.send_json({
                    "type": "alert",
                    "content": "not ur turn"
                })
                return

            index = int(event["index"])
            i , j = index // 3 , index % 3
            if self.board[i][j] == "":

                if not playerChar:
                    await self.disconnect(ws)

                self.board[i][j] = playerChar

                if self.currentPlayer == "X":
                    self.currentPlayer = "O"
                else:
                    self.currentPlayer = "X"

                await self.updateGameState()

                winner = self.checkWinner()
                if winner:
                    # self.isRunning = False
                    if winner != "E":
                        await self.broadcast({
                            "type": "win",
                            "content": winner
                        })
                    else:
                        await self.broadcast({
                            "type": "tie",
                            "content": "no one won"
                        })
                    await self.resetGameState()
                                            

    def checkWinner(self):
        for row in self.board:
            if row[0] != "" and row[0] == row[1] == row[2]:
                return row[0]

        for col in range(3):
            if (
                self.board[0][col] != ""
                and self.board[0][col] == self.board[1][col] == self.board[2][col]
            ):
                return self.board[0][col]

        if (
            self.board[0][0] != ""
            and self.board[0][0] == self.board[1][1] == self.board[2][2]
        ):
            return self.board[0][0]

        if (
            self.board[0][2] != ""
            and self.board[0][2] == self.board[1][1] == self.board[2][0]
        ):
            return self.board[0][2]

        if self.endOfgame():
            return "E"
        return None

    def endOfgame(self):
        for row in self.board:
            for cell in row:
                if cell == "":
                    return False
        return True
    
    async def broadcast(self, message: dict):
        for player in self.players:
            await player["ws"].send_json(message)
    
    async def updateGameState(self):
        await self.broadcast({
            "type": "gameStateUpdate",
            "board": self.board,
            "currentPlayer": self.currentPlayer
        })

    async def resetGameState(self):
        await asyncio.sleep(1)
        print("Game reset")
        self.board = [["" for _ in range(3)] for _ in range(3)]
        self.isRunning = True
        self.currentPlayer = "X"
        await self.updateGameState()


    async def disconnect(self, ws: WebSocket):

        for player in self.players:
            if player["ws"] == ws:
                self.players.remove(player)
                break
                

        print("Player Disconnected")
        print(f"number of players {len(self.players)}")


    def availableChar(self): # informs the player on connect of his char X or O
        if not self.players :
            return "X"
        if self.players[0]["char"] == "X":
            return "O"
        return "X"

    def getPlayerFromWebSocket(self, ws: WebSocket):
        for player in self.players:
            if player["ws"] == ws:
                return player["char"]
        return None

File: frontend/js/game.js
const partyKeyField = document.getElementById("party-key");
const cells = document.getElementsByClassName("game-cell");
let currentPlayer;
let me;
let ws;

const playerChar = document.getElementById("player-char");
const turnStatus = document.getElementById("turn-status");
const gameContainer = document.getElementById("game-container");
const playerName = localStorage.getItem("playerName");

const partyKey = localStorage.getItem("partyKey");
if (!partyKey){
    window.location.href = "index.html";
    
}else{

    console.log("party key = " + partyKey)
    
    const isLocal = window.location.hostname === "127.0.0.1";

    const wsUrl = isLocal
        ? `ws://127.0.0.1:2222/game/ws/${partyKey}/${playerName}`
        : `wss://${window.location.host}/game/ws/${partyKey}/${playerName}`;

    ws = new WebSocket(wsUrl);
    
    
    
    main();
}



function updateTurnStatus(currentPlayer) {
    
    if (!me) {
        return;
    }
    
    playerChar.textContent = me
    
    if (currentPlayer === me) {
        turnStatus.textContent = "Your turn";
        gameContainer.classList.add("my-turn");
        gameContainer.classList.remove("not-my-turn");
    } else {
        turnStatus.textContent = "Opponent's turn";
        gameContainer.classList.add("not-my-turn");
        gameContainer.classList.remove("my-turn");
    }
}

function initGame(){
    console.log("initGame")
    for (const cell of cells) {
        cell.textContent = "";
        cell.addEventListener("click", () => {
            handleCellClick(cell);
        })
    }
    
}

function handleCellClick(cell){
    ws.send(JSON.stringify({
        type: "play",
        index: cell.id,
    }))
}



function main(){
    
    const playerName = localStorage.getItem("playerName");
    if (!playerName){
        window.location.href = "login.html"
        return;
    }
    document.getElementById("player-name").innerHTML += "<mark>" +  playerName + "</mark>"
    
    partyKeyField.value = partyKey

    document.getElementById("copy-party-key-btn").addEventListener("click", async () => {
        try{
            await navigator.clipboard.writeText(partyKey);
            console.log("Clipboard copied text " + partyKey)
            alert("Party key copied");

        }catch (error){
            console.error('Failed to copy text:', error);
        }
    });
    initGame();
    
    ws.addEventListener('open', (event) => {
        console.log('Connected to the Webws server.');
        
        ws.send(JSON.stringify({
            type: "infoMe",
        }))
    })
    ws.addEventListener('message', (event) => {
        const data = JSON.parse(event.data);
    
        if (data.type == "meInfo"){
            me = data.content;
            return
        }
        if (data.type == "gameStateUpdate"){
            console.log("gameStateUpdate")
            updateBoard(data.board) 

            updateTurnStatus(data.currentPlayer);
            return;
        }
        if (data.type == "win") {
            setTimeout(() => {
                alert(data.content + " has won");
            }, 1000);
            return;
        }

        if (data.type == "tie") {
            setTimeout(() => {
                alert(data.content);
            }, 1000);
            return;
        }       
        if (data.type == "alert"){
            alert(data.content);
            return
        }
        if (data.type == "rejection"){
            alert(data.content)
            return;
        }
        if (data.type == "oponent-name"){
            displayOponentName(data.content);
            return;
        }

    });
    
    ws.addEventListener('error', (event) => {
        console.error('Webws error observed:', event);
    });
    
    ws.addEventListener('close', (event) => {
        console.log(`Connection closed. Code: ${event.code}, Reason: ${event.reason}`);
    });
    
}

function updateBoard(board){
    for (let i = 0; i < 3; i++){
        for (let j = 0; j < 3; j++){
            cells[i * 3 + j].textContent = board[i][j]
        }
    }
}

function displayOponentName(names){
    for (const name of names){
        if (name != localStorage.getItem("playerName")){
            document.getElementById("player-name").innerHTML += ` VS <mark>${name}</mark>`
        }
    }
}